kong/router/fields.lua

```diff
@@ -0,0 +1,346 @@
+    -- fallback to preread SNI if current connection doesn't terminate TLS
+
+    FIELDS_FUNCS["tls.sni"] =
+    function(params)
+      if not params.sni then
+        params.sni = server_name() or var.ssl_preread_server_name
+      end
+      return params.sni
+    end
+
+    -- net.*
+    -- when proxying TLS request in second layer or doing TLS passthrough
+    -- rewrite the dst_ip, port back to what specified in proxy_protocol
+
+    FIELDS_FUNCS["net.dst.ip"] =
+    function(params)
+      if not params.dst_ip then
+        if var.kong_tls_passthrough_block == "1" or var.ssl_protocol then
+          params.dst_ip = var.proxy_protocol_server_addr
+        else
+          params.dst_ip = var.server_addr
+        end
+      end
+
+      return params.dst_ip
+    end
+
+    FIELDS_FUNCS["net.dst.port"] =
+    function(params, ctx)
+      if not params.dst_port then
+        if var.kong_tls_passthrough_block == "1" or var.ssl_protocol then
+          params.dst_port = tonumber(var.proxy_protocol_server_port)
+        else
+          params.dst_port = tonumber((ctx or ngx.ctx).host_port, 10) or
+                            tonumber(var.server_port, 10)
+        end
+      end
+
+      return params.dst_port
+    end
+
+end -- is_http
+
+
+if is_http then
+
+  -- func => get_headers or get_uri_args
+  -- name => "headers" or "queries"
+  -- max_config_option => "lua_max_req_headers" or "lua_max_uri_args"
+  local function get_http_params(func, name, max_config_option)
+    local params, err = func()
+    if err == "truncated" then
+      local max = kong and kong.configuration and kong.configuration[max_config_option] or 100
+      ngx.log(ngx.ERR,
+              string.format("router: not all request %s were read in order to determine the route " ..
+                            "as the request contains more than %d %s, " ..
+                            "route selection may be inaccurate, " ..
+                            "consider increasing the '%s' configuration value " ..
+                            "(currently at %d)",
+                            name, max, name, max_config_option, max))
+    end
+
+    return params
+  end
+
+
+  setmetatable(FIELDS_FUNCS, {
+  __index = function(_, field)
+    local func
+
+    local prefix = field:sub(1, PREFIX_LEN)
+
+    if prefix == HTTP_HEADERS_PREFIX then
+      func = function(params)
+        if not params.headers then
+          params.headers = get_http_params(get_headers, "headers", "lua_max_req_headers")
+        end
+
+        return params.headers[field:sub(PREFIX_LEN + 1)]
+      end
+
+    elseif prefix == HTTP_QUERIES_PREFIX then
+      func = function(params)
+        if not params.queries then
+          params.queries = get_http_params(get_uri_args, "queries", "lua_max_uri_args")
+        end
+
+        return params.queries[field:sub(PREFIX_LEN + 1)]
+      end
+    end
+
+    -- cache the lazily generated method
+    FIELDS_FUNCS[field] = func
```

Comment:
I don't think we should cache this, as the number of header name could be infinite and this table may blow up. Just returning `func` here now should be enough.

Reply:
done.